tempodb/encoding/vparquet/block_search_tags.go

```diff
@@ -0,0 +1,312 @@
+	if err != nil {
+		return fmt.Errorf("unexpected error opening parquet file: %w", err)
+	}
+	defer func() { span.SetTag("inspectedBytes", rr.TotalBytesRead.Load()) }()
+
+	return searchTags(derivedCtx, cb, pf)
+}
+
+func searchTags(_ context.Context, cb common.TagCallback, pf *parquet.File) error {
+	// find indexes of generic attribute columns
+	resourceKeyIdx, _ := pq.GetColumnIndexByPath(pf, FieldResourceAttrKey)
+	spanKeyIdx, _ := pq.GetColumnIndexByPath(pf, FieldSpanAttrKey)
+	if resourceKeyIdx == -1 || spanKeyIdx == -1 {
+		return fmt.Errorf("resource or span attributes col not found (%d, %d)", resourceKeyIdx, spanKeyIdx)
+	}
+	standardAttrIdxs := []int{
+		resourceKeyIdx,
+		spanKeyIdx,
+	}
+
+	// find indexes of all special columns
+	specialAttrIdxs := map[int]string{}
+	for lbl, col := range labelMappings {
+		idx, _ := pq.GetColumnIndexByPath(pf, col)
+		if idx == -1 {
+			continue
+		}
+
+		specialAttrIdxs[idx] = lbl
+	}
+
+	// now search all row groups
+	var err error
+	rgs := pf.RowGroups()
+	for _, rg := range rgs {
+		// search all special attributes
+		for idx, lbl := range specialAttrIdxs {
+			cc := rg.ColumnChunks()[idx]
+			err = func() error {
+				pgs := cc.Pages()
+				defer pgs.Close()
+				for {
+					pg, err := pgs.ReadPage()
+					if err == io.EOF || pg == nil {
+						break
+					}
+					if err != nil {
+						return err
+					}
+
+					stop := func(page parquet.Page) bool {
+						defer parquet.Release(page)
+
+						// if a special attribute has any non-null values, include it
+						if page.NumNulls() < page.NumValues() {
+							cb(lbl)
+							delete(specialAttrIdxs, idx) // remove from map so we won't search again
+							return true
+						}
+						return false
+					}(pg)
+					if stop {
+						break
+					}
+				}
+				return nil
+			}()
+			if err != nil {
+				return err
+			}
+		}
+
+		// search other attributes
+		for _, idx := range standardAttrIdxs {
+			cc := rg.ColumnChunks()[idx]
+			err = func() error {
+				pgs := cc.Pages()
+				defer pgs.Close()
+				for {
+					pg, err := pgs.ReadPage()
+					if err == io.EOF || pg == nil {
+						break
+					}
+					if err != nil {
+						return err
+					}
+
+					func(page parquet.Page) {
+						defer parquet.Release(page)
+
+						dict := page.Dictionary()
+						if dict == nil {
+							return
+						}
+
+						for i := 0; i < dict.Len(); i++ {
+							s := dict.Index(int32(i)).String()
+							cb(s)
+						}
+					}(pg)
+				}
+				return nil
+			}()
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (b *backendBlock) SearchTagValues(ctx context.Context, tag string, cb common.TagCallback, opts common.SearchOptions) error {
+
+	att, ok := translateTagToAttribute[tag]
+	if !ok {
+		att = traceql.NewAttribute(tag)
+	}
+
+	// Wrap to v2-style
+	cb2 := func(v traceql.Static) bool {
+		cb(strings.Trim(v.String(), "`"))
+		return false
+	}
+
+	return b.SearchTagValuesV2(ctx, att, cb2, opts)
```

Comment:
SearchTagsV1 is now internally a wrapper around the V2 logic. This is good overall but led to a few quirky bits because the root span-name/service-name columns are addressable in V1 (ex: `/search/tag/root.name/values`) but technically don't exist in TraceQL.   So the `nonTraceQLAttributes` map catches these and is checked below.